notifications: Don't enqueue notifications for bots.

abhijeetbodas2001 opened the pull request:
This replaces the temporary (and testless) fix in
24b1439 with a more permanent
fix.

Instead of checking if the user is a bot just before
sending the notifications, we now just don't enqueue
notifications for bots. This is done by sending a list
of bot IDs to the event_queue code, just like other
lists which are used for creating NotificationData objects.

abhijeetbodas2001 commented:
@timabbott FYI :slightly_smiling_face:

abhijeetbodas2001 commented on zerver/lib/actions.py:
```diff
@@ -1620,15 +1620,11 @@ def should_send(setting: str, row: Dict[str, Any]) -> bool:
-            # determining whether this wildcard mention should be
-            # treated as a mention (and follow the user's mention
-            # notification preferences) or a normal message.
+            # determine the set users have turned on the "wildcard
```
Ok, updated :+1:

andersk commented on zerver/tests/test_notification_data.py:
```diff
@@ -212,6 +216,43 @@ def test_is_notifiable(self) -> None:
         self.assertTrue(user_data.is_notifiable(acting_user_id=acting_user_id, idle=True))
 
+    def test_bot_user_notifiability(self) -> None:
+        class function_arguments(TypedDict):
```
Here’s the [developer community thread](https://chat.zulip.org/#narrow/stream/92-learning/topic/mypy.3A.20Reusing.20arguments.20for.20function.20calls/near/1299681) where you asked about reducing the verbosity of this `TypedDict` construction. I think a [wrapper function](https://chat.zulip.org/#narrow/stream/92-learning/topic/mypy.3A.20Reusing.20arguments.20for.20function.20calls/near/1300149) or a [`for` loop](https://chat.zulip.org/#narrow/stream/92-learning/topic/mypy.3A.20Reusing.20arguments.20for.20function.20calls/near/1300561) would be a simpler solution.

abhijeetbodas2001 replied:
Thanks. I liked the `for` loop approach, and have pushed the same here.

timabbott commented:
I merged the first two commits as the series ending with aa18e797a8a9a53bb32e73880b58831274a349b8, after tweaking the language in the updated comment.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -569,9 +569,6 @@ def handle_missedmessage_emails(
-        return
 
     if not user_profile.enable_offline_email_notifications:
         # BUG: Investigate why it's possible to get here.
```
The theory for how we'd get here is just that the user configured notifications for the target stream, right?  Perhaps we can turn this into a check verifying that the user has configured notifications for the stream, and if not, log an error and return?  We can probably delete that code once it's been in production and confirmed our understanding of the world, but that'd let us fully close out this investigation.

abhijeetbodas2001 replied:
> The theory for how we'd get here is just that the user configured notifications for the target stream, right?

You mean the block with the `# BUG: Investigate why it's possible to get here.`, right? That shall be handled as part of fixing #19451. I'm only working only the "bots" case in this PR.

abhijeetbodas2001 replied:
I expect #19451 to be bunch of work. Not sure when I'll have the time for it, but I'm going to take that up next.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -569,9 +569,6 @@ def handle_missedmessage_emails(
     user_profile = get_user_profile_by_id(user_profile_id)
-    if user_profile.is_bot:  # nocoverage # TODO -- needs a test.
-        # Never email bot users.
-        return
```
Do we want an `assert not user_profile.is_bot` here?  it's a bug if we ever get here for a bot user, and worth writing that down.

Though I suppose it's possible to have send-email events in the queue at the time the server restarts, and we wouldn't want those to cause this to error, so maybe the right model is a `logger.error` + return here?  

@alexmv for thoughts.

alexmv replied:
I'd log an error and return, yeah.

abhijeetbodas2001 replied:
> Though I suppose it's possible to have send-email events in the queue at the time the server restarts, and we wouldn't want those to cause this to error, so maybe the right model is a logger.error + return here?

OK, makes sense. Added the logging code, along with a `TODO/compatibility`.

alexmv commented on zerver/lib/actions.py:
```diff
@@ -1620,15 +1620,11 @@ def should_send(setting: str, row: Dict[str, Any]) -> bool:
-            # determining whether this wildcard mention should be
-            # treated as a mention (and follow the user's mention
-            # notification preferences) or a normal message.
+            # determine the set users have enabled the "wildcard
```
```suggestion
            # determine the set of users who have enabled the "wildcard
```

abhijeetbodas2001 replied:
I believe Tim fixed this while merging acdce4df4797171fbcd807e0d39fc94d48e475dd; marking this as resolved. Thanks!

alexmv commented on zerver/lib/actions.py:
```diff
@@ -1731,6 +1732,10 @@ def is_service_bot(row: Dict[str, Any]) -> bool:
 
     service_bot_tuples = [(row["id"], row["bot_type"]) for row in rows if is_service_bot(row)]
 
+    # We also need the user IDs of all bots, to avoid trying to send push/email
```
Do we want to update the comment on line 1719 which talks about "These **two** bot data structures..." now that there are three of them?

abhijeetbodas2001 replied:
Yes, thanks for pointing out. Done.

alexmv commented on zerver/tests/test_notification_data.py:
```diff
@@ -212,6 +212,24 @@ def test_is_notifiable(self) -> None:
 
+    def test_bot_user_notifiability(self) -> None:
+        # Non-bot user (`user_id=9`) should get notified, while bot user (`user_id=10`) shouldn't
+        for user_id, notifiable in [(9, True), (10, False)]:
```
These IDs seem rather fragile.  Can we iterate over email addresses, and look up their IDs in the loop?

alexmv commented on zerver/tests/test_notification_data.py:
```diff
@@ -212,6 +212,24 @@ def test_is_notifiable(self) -> None:
+                online_push_user_ids=set(),
+                pm_mention_email_disabled_user_ids=set(),
+                pm_mention_push_disabled_user_ids=set(),
+                all_bot_user_ids={10, 11, 12},
```
This list of IDs also seems fragile.  Can we pull the same trick of hard-coding email addresses, and convert those to IDs for this set?

abhijeetbodas2001 commented:
@alexmv @timabbott Thanks for the reviews! I've updated the PR after addressing the comments.

@alexmv The user IDs aren't fragile, because they don't really represent any users in the test database. The class itself does not do any database lookups etc. In some sense, the user IDs are nothing more than just plain integers for our class, and the test verifies how the class/method behaves on those integers.

alexmv commented:
> The user IDs aren't fragile, because they don't really represent any users in the test database. The class itself does not do any database lookups etc. In some sense, the user IDs are nothing more than just plain integers for our class, and the test verifies how the class/method behaves on those integers.

Ah, thanks for clarifying!  Maybe a comment in the test that these IDs are artificial and arbitrary, and merely chosen with respect to the value passed as `all_bot_user_ids`, would clarify for anyone else that might be similarly concerned.

Otherwise, looks good to me.  I wonder a little, looking at it again, if `error` is the right level to log at when we see a bot notification.  If it's possible to get it in normal behaviour during upgrade, and there's no action for the server admin to take in those contexts, `warning` might be better?

timabbott commented:
I think it's unlikely in practice that one would see this during an upgrade, but possible.  `warning` is probably OK for our goal here.

Thinking out loud a bit, I wonder if we should add some code to include some representation of the Zulip version details in events that are put into our queue, so that one can do error-level logging `if event['zulip_version'] != current_zulip_version`, and otherwise log nothing?

abhijeetbodas2001 commented:
I haven't worked in that part, but is there a way for the server to know when it starts that is was upgraded?

Or, is there a way to, as a standard-operating-procedure, process existing events in the queue and only then start the upgrade? Just wondering if/how we could avoid this problem all together.

alexmv commented:
Whether to ensure the queues are all empty before starting back up is a downtime performance tradeoff.  The only way to not have to serve old queue data with new code is to stop web traffic, wait for all workers to clear the queues, and then restart the workers and django with new code.  Which means that we'd be _required_ to have a window where we're not serving http traffic, with a gap we don't fully control.  Which means throwing rolling django restarts out of the window.

Actually, having written that, I think it's not quite true.  There are other ways to do it, but they get probably _more_ complicated than making things back-compat in the queue workers.  For instance, we could encode the version into the queue name itself -- so upgrading would be:
1. start new workers, watching new queue names
2. restart django, which stops writing to old queue names, and starts writing to new queue names
3. wait for old queue workers to clear their queues before shutting them down

But the failure mode of this is that it gets much easier to accidentally drop events, if the old queue workers die for any reason. There may also be dependencies between queues which mean you have to wait for everything to zero before stopping any of them -- for instance, even if the `embed_links` queue is empty, if `email_mirror` is not yet, an incoming email may send a message which enqueues an `embed_links` (with the _old_ version queue name, because the `email_mirror` worker is still running the old code).  All of this while ensuring somehow via supervisor that these old versions of the workers are able to stay up and running until they're no longer needed.

There is complexity to being back-compat within the worker processes, but I think it far outweighs the operational complexity of managing full executors of the old code.

alexmv commented:
Popping the stack to the specific instance here, I'm +1 to merge this with just a warning and return; I don't think we should block on anything more complicated here.

abhijeetbodas2001 commented:
Ok, pushed with those changes (`error` -> `warn`). I also added a comment addressing this:

> Maybe a comment in the test that these IDs are artificial and arbitrary, and merely chosen with respect to the value passed as all_bot_user_ids, would clarify for anyone else that might be similarly concerned.

to the test class itself, since other tests in that file similarly use arbitrary user IDs.

alexmv commented:
Thanks for humoring me on the docstring!  This looks great, @abhijeetbodas2001, and I'm happy to have our random test-less return made into a more robust system.  Merged as 15e871784725a817499e74119e57838487ec5186.
